File: opensquilla-webui/src/tauri/events.ts
/**
 * Tauri event listener manager.
 *
 * In the WebSocket world the frontend subscribed to gateway events through
 * `rpc.on('session.event.text_delta', handler)` (see `src/lib/rpc.ts`). Under
 * Tauri the same events are emitted from Rust via `app_handle.emit(event, payload)`
 * and consumed in JS with `listen(event, handler)` from `@tauri-apps/api/event`.
 *
 * The Rust runtime emits these channels:
 *
 *   - `agent:stream:{sessionId}` — per-session turn/stream/tool events. The
 *     payload is `MessageStreamEvent` (see `src-tauri/src/ipc.rs`), a serde
 *     internally-tagged enum (`#[serde(tag="type", content="data")]`) whose
 *     `type` discriminator is one of `turn_start | generation_start |
 *     stream_event | tool_call_start | tool_call_complete | turn_complete |
 *     turn_error | compaction`. The `stream_event` data is itself an
 *     internally-tagged enum (`#[serde(tag="event", content="data")]`) whose
 *     `event` discriminator is `content_block_start | content_block_delta |
 *     content_block_stop | message_delta | message_stop | error | ping`.
 *     Field names inside `data` are snake_case (these enums are NOT
 *     `rename_all = "camelCase"`), while nested shared structs that are
 *     camelCase-renamed (e.g. `UsagePayload`) keep their camelCase fields.
 *   - `sessions:list-changed` — session list refresh (payload null).
 *   - `gateway:status` — gateway lifecycle changes (`{ running, url?, port?, error? }`).
 *   - `locale://changed` — locale override persisted (`{ locale }`).
 *   - `update://state` — updater state snapshot.
 *
 * This module provides:
 *  - {@link TauriEventListener} / {@link UnlistenFn}: transport-agnostic types.
 *  - {@link listen}: a low-level listener that talks to `window.__TAURI__.event`
 *    when present and no-ops (returning a NOP unlisten) when not, so dev mode
 *    without Tauri never crashes.
 *  - {@link useTauriEvent}: a Vue composable that binds a listener to the
 *    component lifecycle and auto-cleans up on unmount, mirroring
 *    `useRpcEvent` in `src/composables/useRpc.ts`.
 *  - {@link useTauriStreamEvents}: a Vue composable that subscribes to the
 *    per-session `agent:stream:{sessionId}` channel and dispatches each
 *    `MessageStreamEvent` to the turn / stream / tool handler groups by its
 *    `type` tag.
 */

import { onUnmounted } from 'vue'
import { isTauri, TauriInvokeError } from './invoke'

/** A function returned by `listen` that detaches the handler. */
export type UnlistenFn = () => void

/** A handler invoked for every matching Tauri event. */
export type TauriEventListener<T = unknown> = (payload: T) => void

/** Shape of the Tauri event API injected by the shell (@tauri-apps/api v2). */
export interface TauriEventApi {
  listen<T = unknown>(
    event: string,
    handler: (e: { event: string; id: number; payload: T }) => void,
  ): Promise<UnlistenFn>
}

export interface TauriEventGlobal {
  __TAURI__?: {
    event?: TauriEventApi
    core?: unknown
  }
}

/* ──────────────────────────────────────────────────────────────────────────
 * Wire payload types. These mirror the serde enums/structs in
 * `src-tauri/src/ipc.rs` exactly. `MessageStreamEvent` and `StreamEventPayload`
 * are internally tagged (`type` / `event`) with `data` content; their fields
 * are snake_case because those enums have no `rename_all`.
 * ────────────────────────────────────────────────────────────────────────── */

/** `ContentBlockDto` — a single content block inside a `MessageDto`. */
export interface ContentBlockDto {
  type: 'text' | 'tool_use' | 'tool_result' | 'reasoning' | (string & {})
  /** text blocks */
  text?: string
  /** tool_use blocks */
  id?: string
  name?: string
  input?: unknown
  /** tool_result blocks */
  tool_use_id?: string
  content?: string
  is_error?: boolean
  /** reasoning blocks */
  reasoning?: string
}

/** `MessageDto` — a message in a chat history or turn_complete payload. */
export interface MessageDto {
  role: string
  content: ContentBlockDto[]
  name?: string
  tool_call_id?: string
}

/** `UsagePayload` — camelCase-renamed token usage. */
export interface UsagePayload {
  inputTokens: number
  outputTokens: number
  totalTokens: number
}

/** `StreamDeltaPayload` — internally tagged stream delta. */
export type StreamDeltaPayload =
  | { type: 'text_delta'; text: string }
  | { type: 'reasoning_delta'; reasoning: string }
  | { type: 'input_json_delta'; partial_json: string }

/**
 * `StreamEventPayload` — the inner payload carried by a `stream_event`
 * `MessageStreamEvent`. Internally tagged with `event` and `data`.
 */
export type StreamEventPayload =
  | { event: 'content_block_start'; data: { index: number; block: ContentBlockDto } }
  | { event: 'content_block_delta'; data: { index: number; delta: StreamDeltaPayload } }
  | { event: 'content_block_stop'; data: { index: number } }
  | {
      event: 'message_delta'
      data: { stop_reason?: string | null; usage?: UsagePayload | null }
    }
  | { event: 'message_stop'; data: { content: ContentBlockDto[]; usage?: UsagePayload | null } }
  | { event: 'error'; data: { message: string; code?: string | null } }
  | { event: 'ping'; data?: null }

/**
 * `MessageStreamEvent` — the payload emitted on `agent:stream:{sessionId}`.
 * Internally tagged with `type` and `data`.
 */
export type MessageStreamEvent =
  | { type: 'turn_start'; data: { turn_id: string; session_id: string; timestamp: string } }
  | { type: 'generation_start'; data: { model: string; provider: string } }
  | { type: 'stream_event'; data: StreamEventPayload }
  | {
      type: 'tool_call_start'
      data: { tool_call_id: string; name: string; input: unknown }
    }
  | {
      type: 'tool_call_complete'
      data: {
        tool_call_id: string
        name: string
        content: string
        is_error: boolean
        duration_ms: number
      }
    }
  | {
      type: 'turn_complete'
      data: {
        turn_id: string
        session_id: string
        messages: MessageDto[]
        usage: UsagePayload
        duration_ms: number
      }
    }
  | {
      type: 'turn_error'
      data: { turn_id: string; session_id: string; message: string; code?: string | null }
    }
  | { type: 'compaction'; data: { before_count: number; after_count: number; success: boolean } }

/** The full channel name for a session's streaming events. */
export function agentStreamChannel(sessionId: string): string {
  return `agent:stream:${sessionId}`
}

/** `GatewayStatusEvent` — payload of the `gateway:status` channel. */
export interface GatewayStatusEvent {
  running: boolean
  url?: string | null
  port?: number | null
  error?: string | null
}

/** Payload of the `locale://changed` channel. */
export interface LocaleChangedEvent {
  locale: string
}

/** Payload of the `update://state` channel. */
export interface UpdateStateEvent {
  status: string
  available?: boolean
  version?: string | null
  release_url?: string | null
  error?: string | null
  downloading?: boolean
  downloaded?: boolean
  applying?: boolean
  [key: string]: unknown
}

/**
 * Mapping of the top-level Tauri event channel names to their payload types.
 * The streaming channel `agent:stream:{sessionId}` is session-scoped, so it is
 * not listed here (use {@link agentStreamChannel} + {@link listen} or
 * {@link useTauriStreamEvents}).
 */
export interface TauriEventMap {
  'sessions:list-changed': null
  'gateway:status': GatewayStatusEvent
  'locale://changed': LocaleChangedEvent
  'update://state': UpdateStateEvent
}

/** Union of all top-level Tauri event names. */
export type TauriEventName = keyof TauriEventMap

/** A Tauri event name OR an arbitrary string (for one-off listeners). */
export type EventName = TauriEventName | (string & {})

function getTauriEventApi(): TauriEventApi | null {
  const tauri = (globalThis as unknown as TauriEventGlobal).__TAURI__
  return tauri?.event ?? null
}

const nopUnlisten: UnlistenFn = () => {}

/**
 * Low-level Tauri event listener. Subscribes to `eventName` and invokes
 * `handler` for every matching emission. Returns an unlisten function.
 *
 * When not running in a Tauri shell (dev server, vitest), this returns a NOP
 * unlistener instead of throwing, so components and adapters can call it
 * unconditionally. Callers that need to know whether the subscription is live
 * should check {@link isTauri} first.
 */
export function listen<T = unknown>(
  eventName: EventName,
  handler: TauriEventListener<T>,
): UnlistenFn {
  const api = getTauriEventApi()
  if (!api) {
    return nopUnlisten
  }
  let active = true
  let detach: UnlistenFn | null = null
  // `listen` returns a Promise<UnlistenFn>; we capture it so a fast unmount
  // before resolution can still tear down once it lands.
  api.listen<T>(eventName, (e) => {
    if (!active) return
    handler(e.payload)
  })
    .then((un) => {
      if (!active) {
        // Component already unmounted before the listener attached.
        try { un() } catch { /* ignore */ }
        return
      }
      detach = un
    })
    .catch(() => {
      // A failed listen (e.g. event plugin missing on older shells) is
      // non-fatal; the caller simply won't receive events.
    })
  return () => {
    if (!active) return
    active = false
    if (detach) {
      try { detach() } catch { /* ignore */ }
      detach = null
    }
  }
}

/**
 * Vue composable: subscribe to a Tauri event for the lifetime of the calling
 * component and auto-cleanup on unmount. Mirrors `useRpcEvent` from
 * `src/composables/useRpc.ts`, so chat composables can swap transports with a
 * one-line change.
 *
 * Must be called from `setup()` (relies on `onUnmounted`).
 *
 * @example
 *   useTauriEvent('gateway:status', (p: GatewayStatusEvent) => {
 *     if (p.running) setGatewayUrl(p.url)
 *   })
 */
export function useTauriEvent<E extends TauriEventName>(
  eventName: E,
  handler: TauriEventListener<TauriEventMap[E]>,
): { unlisten: UnlistenFn }

export function useTauriEvent<T = unknown>(
  eventName: EventName,
  handler: TauriEventListener<T>,
): { unlisten: UnlistenFn }

export function useTauriEvent<T = unknown>(
  eventName: EventName,
  handler: TauriEventListener<T>,
): { unlisten: UnlistenFn } {
  const unlisten = listen<T>(eventName, handler)
  onUnmounted(() => {
    unlisten()
  })
  return { unlisten }
}

/**
 * Handler groups for the per-session streaming channel. Each `MessageStreamEvent`
 * is routed by its `type` tag:
 *
 *   turn_start | generation_start | turn_complete | turn_error | compaction → onTurnEvent
 *   stream_event                                                          → onStreamEvent
 *   tool_call_start | tool_call_complete                                  → onToolEvent
 */
export interface StreamHandlers {
  onTurnEvent?: TauriEventListener<MessageStreamEvent>
  onStreamEvent?: TauriEventListener<StreamEventPayload>
  onToolEvent?: TauriEventListener<MessageStreamEvent>
}

/**
 * Subscribe to a session's `agent:stream:{sessionId}` channel and dispatch to
 * the turn / stream / tool handler groups by the `MessageStreamEvent.type` tag.
 * Returns a single unlisten that tears down the underlying Tauri listener.
 *
 * Must be called from `setup()` (relies on `onUnmounted`). The listen should be
 * attached BEFORE `send_message` is invoked so no early turn events are missed.
 */
export function useTauriStreamEvents(
  sessionId: string,
  handlers: StreamHandlers,
): { unlisten: UnlistenFn } {
  const unlisten = listen<MessageStreamEvent>(agentStreamChannel(sessionId), (payload) => {
    if (!payload || typeof payload.type !== 'string') return
    switch (payload.type) {
      case 'stream_event':
        handlers.onStreamEvent?.(payload.data)
        break
      case 'tool_call_start':
      case 'tool_call_complete':
        handlers.onToolEvent?.(payload)
        break
      case 'turn_start':
      case 'generation_start':
      case 'turn_complete':
      case 'turn_error':
      case 'compaction':
      default:
        handlers.onTurnEvent?.(payload)
        break
    }
  })
  onUnmounted(unlisten)
  return { unlisten }
}

/** Re-exported so adapters can pattern-match errors from listen setup. */
export { isTauri, TauriInvokeError }

File: crates/core/src/events.rs
use crate::types::{ContentBlock, Message, ToolCall, ToolResult, Usage};
use serde::{Deserialize, Serialize};

/// Events emitted during a streaming model response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", content = "data")]
pub enum StreamEvent {
    /// A new content block has started.
    #[serde(rename = "content_block_start")]
    ContentBlockStart {
        /// The index of the content block.
        index: usize,
        /// The content block that started.
        block: ContentBlock,
    },
    /// A delta update to an existing content block.
    #[serde(rename = "content_block_delta")]
    ContentBlockDelta {
        /// The index of the content block being updated.
        index: usize,
        /// The delta update to apply.
        delta: ContentBlockDelta,
    },
    /// A content block has been completed.
    #[serde(rename = "content_block_stop")]
    ContentBlockStop {
        /// The index of the content block that stopped.
        index: usize,
    },
    /// A message delta update (e.g., stop reason change).
    #[serde(rename = "message_delta")]
    MessageDelta {
        /// The delta to apply to the message.
        delta: MessageDelta,
        /// Updated usage information.
        usage: Option<Usage>,
    },
    /// The message has been fully streamed and is complete.
    #[serde(rename = "message_stop")]
    MessageStop {
        /// The complete message content blocks.
        content: Vec<ContentBlock>,
        /// Final usage information.
        usage: Option<Usage>,
    },
    /// An error occurred during streaming.
    #[serde(rename = "error")]
    Error {
        /// The error message.
        message: String,
        /// Optional error code.
        code: Option<String>,
    },
    /// A ping keepalive event.
    #[serde(rename = "ping")]
    Ping,
}

/// A delta update to a content block during streaming.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ContentBlockDelta {
    /// A text delta: partial text content.
    #[serde(rename = "text_delta")]
    TextDelta {
        /// The partial text content.
        text: String,
    },
    /// A reasoning delta: partial reasoning content.
    #[serde(rename = "reasoning_delta")]
    ReasoningDelta {
        /// The partial reasoning content.
        reasoning: String,
    },
    /// A tool call delta: partial tool call arguments.
    #[serde(rename = "input_json_delta")]
    InputJsonDelta {
        /// The partial JSON input string.
        partial_json: String,
    },
}

/// A delta update to a message during streaming.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageDelta {
    /// The stop reason for the message, if changed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop_reason: Option<String>,
    /// The stop sequence that terminated generation, if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop_sequence: Option<String>,
}

/// Events emitted during a conversation turn.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", content = "data")]
pub enum TurnEvent {
    /// The turn has started.
    #[serde(rename = "turn_start")]
    TurnStart {
        /// The turn identifier.
        turn_id: String,
        /// The timestamp when the turn started.
        timestamp: String,
    },
    /// A message has been received from the user.
    #[serde(rename = "user_message")]
    UserMessage {
        /// The user message.
        message: Message,
    },
    /// The model has started generating a response.
    #[serde(rename = "generation_start")]
    GenerationStart {
        /// The model being used.
        model: String,
        /// The provider serving the request.
        provider: String,
    },
    /// A stream event from the model response.
    #[serde(rename = "stream_event")]
    StreamEvent {
        /// The stream event from the model.
        event: StreamEvent,
    },
    /// The turn has completed successfully.
    #[serde(rename = "turn_complete")]
    TurnComplete {
        /// The final assistant message.
        message: Message,
        /// Token usage for this turn.
        usage: Usage,
        /// Duration of the turn in milliseconds.
        duration_ms: u64,
    },
    /// The turn encountered an error.
    #[serde(rename = "turn_error")]
    TurnError {
        /// The error message.
        message: String,
        /// Optional error code.
        code: Option<String>,
    },
    /// Context compaction has been triggered.
    #[serde(rename = "compaction")]
    Compaction {
        /// Number of messages before compaction.
        before_count: usize,
        /// Number of messages after compaction.
        after_count: usize,
        /// Whether compaction was successful.
        success: bool,
    },
}

/// Events emitted during tool execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", content = "data")]
pub enum ToolEvent {
    /// A tool call has been initiated.
    #[serde(rename = "tool_call_start")]
    ToolCallStart {
        /// The tool call details.
        call: ToolCall,
        /// The timestamp when the tool call started.
        timestamp: String,
    },
    /// A tool call has completed successfully.
    #[serde(rename = "tool_call_complete")]
    ToolCallComplete {
        /// The tool call that was executed.
        call: ToolCall,
        /// The result of the tool execution.
        result: ToolResult,
        /// Duration of the tool execution in milliseconds.
        duration_ms: u64,
    },
    /// A tool call has failed.
    #[serde(rename = "tool_call_error")]
    ToolCallError {
        /// The tool call that failed.
        call: ToolCall,
        /// The error message.
        error: String,
        /// Duration of the failed attempt in milliseconds.
        duration_ms: u64,
    },
    /// Multiple tool calls are being executed in parallel.
    #[serde(rename = "tool_parallel_start")]
    ToolParallelStart {
        /// The tool calls being executed.
        calls: Vec<ToolCall>,
    },
    /// All parallel tool calls have completed.
    #[serde(rename = "tool_parallel_complete")]
    ToolParallelComplete {
        /// The results of all tool executions.
        results: Vec<ToolResult>,
        /// Total duration in milliseconds.
        total_duration_ms: u64,
    },
    /// A tool call has been rejected (e.g., due to safety checks).
    #[serde(rename = "tool_rejected")]
    ToolRejected {
        /// The tool call that was rejected.
        call: ToolCall,
        /// The reason for rejection.
        reason: String,
    },
}

/// Combine multiple event types into a single observable event.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum GatewayEvent {
    /// A stream event from model generation.
    Stream(StreamEvent),
    /// A turn lifecycle event.
    Turn(TurnEvent),
    /// A tool execution event.
    Tool(ToolEvent),
}
